challenges/cloud/sre.md

```diff
@@ -0,0 +1,376 @@
+* Ability to run the server
+* Ability to execute integration tests against the local Kubernetes cluster
+
+## Level 3
+
+### Server
+
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
+* HTTP API to set the replica count of a the Kubernetes Deployment
+* HTTP API to get the list of available Deployments in the Kubernetes cluster
+* HTTP health check verifying Kubernetes connectivity
+
+### Automation
+
+* Ability to build the server in a Docker container
+* Ability to run the server
+* Ability to execute integration tests against the local Kubernetes cluster
+* GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass
+
+### Deployment
+
+* Create a helm chart for the service
+
+## Level 4
+
+### Design
+
+In addition to including details about the requirements below, the design
+document should include a section describing how this solution could be deployed
+to an AWS EKS cluster.
+
+### Server
+
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
+* HTTP API to set the replica count of a the Kubernetes Deployment
+* HTTP API to get the list of available Deployments in the Kubernetes cluster
+* HTTP health check verifying Kubernetes connectivity
+* Extend the API to support a difference between current and desired state
+* Extend the Server to support the ability to show the differences between current and desired state
+* Secure connections with mTLS
+
+### Automation
+
+* Ability to build the server in a Docker container
+* Ability to run the server
+* Ability to execute integration tests against the local Kubernetes cluster
+* GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass
+* Produce production ready releases (binaries and docker image)
+
+### Deployment
+
+* Create a helm chart for the service
+* Deploy releases of the API Server to a Kubernetes cluster (and possible dependencies)
+* Upgrade releases of the API Server to a Kubernetes cluster (and possible dependencies)
+* Include a production Deployment of this service, including but not limited to: Deployment, ServiceAccount, Service
+
+
+## Level 5
+
+### Design
+
+In addition to including details about the requirements below, the design
+document should include a section describing how this solution could be deployed
+to an AWS EKS cluster. Please include security considerations as well.
+
+### Server
+
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
+* HTTP API to set the replica count of a the Kubernetes Deployment
+* HTTP API to get the list of available Deployments in the Kubernetes cluster
+* HTTP health check verifying Kubernetes connectivity
+* Extend the API to support a difference between current and desired state
+* Extend the Server to support the ability to show the differences between current and desired state
+* Secure connections with mTLS
+* Replace the HTTP API with gRPC
```

Comment:
In our challenges for Core, switching from. HTTP to gRPC was (IIRC) a level 3 item.

For SRE, maybe this makes sense to show that software engineering skills are more important at higher levels, but I also question whether this is just extra busy work that doesn't really differentiate level 4 and 5 candidates in any dimension other than how much time they have to work on it.

Reply:
> For SRE, maybe this makes sense to show that software engineering skills are more important at higher levels

This was a core part of why this was proposed, but there are other signals that I thought were important:

- securing via mTLS is expected to be more challenging, and can be more brittle compared to a web server
- We don't provide much in the way of a skeleton for gRPC, so it needs to be implemented from the ground up, rather than adapting what is already there
- how the protobuf contract is proposed is an interesting signal to me, it's more nuanced than HTTP, especially making a protobuf schema future proof 

In a nutshell, I'd probably make the distinguish the difference between the two as: An L4 candidate is expected to augment and adapt an existing codebase, whereas an L5 is expected to innovate and challenge the status quo.

With that said, I'm not super firm on getting these signals from gRPC -- though I do think it mirrors our working environment quite nicely

Reply:
> In a nutshell, I'd probably make the distinguish the difference between the two as: An L4 candidate is expected to augment and adapt an existing codebase, whereas an L5 is expected to innovate and challenge the status quo.

I think that's describing it very well, exactly.